File: internal/domain/commits.go
package domain

import (
	"regexp"
	"sort"
)

var padraoChamado = regexp.MustCompile(`\bch(\d+)\b`)
var padraoVB = regexp.MustCompile(`\b(VB-\d+)\b`)

// MatchExato filtra candidatos de grep por word-boundary - ch<num> e VB-<num>
// exatos, nao substring (§4 "Precisao do match"). search_commits do GitRepo so
// traz candidatos brutos.
func MatchExato(candidatos []CommitRef, chamado, vbID string) []CommitRef {
	var padroes []*regexp.Regexp
	if chamado != "" {
		padroes = append(padroes, regexp.MustCompile(`\bch`+regexp.QuoteMeta(chamado)+`\b`))
	}
	if vbID != "" {
		padroes = append(padroes, regexp.MustCompile(`\b`+regexp.QuoteMeta(vbID)+`\b`))
	}

	var resultado []CommitRef
	for _, c := range candidatos {
		for _, p := range padroes {
			if p.MatchString(c.Msg) {
				resultado = append(resultado, c)
				break
			}
		}
	}
	return resultado
}

// ExtrairChamado acha um numero de chamado (ch<num>) na mensagem, usado na
// reconstrucao do lock (§3) para reagrupar por chamado a partir do trailer.
func ExtrairChamado(msg string) (chamado string, ok bool) {
	m := padraoChamado.FindStringSubmatch(msg)
	if m == nil {
		return "", false
	}
	return m[1], true
}

// ExtrairVBID acha um id VB-<num> na mensagem.
func ExtrairVBID(msg string) (vbID string, ok bool) {
	m := padraoVB.FindStringSubmatch(msg)
	if m == nil {
		return "", false
	}
	return m[1], true
}

// OrdenarPorData ordena por CommitDate asc - nao depende de flag do git (§5
// "Ordenacao"). Nao muta a fatia de entrada.
func OrdenarPorData(commits []CommitRef) []CommitRef {
	ordenado := make([]CommitRef, len(commits))
	copy(ordenado, commits)
	sort.Slice(ordenado, func(i, j int) bool {
		return ordenado[i].CommitDate.Before(ordenado[j].CommitDate)
	})
	return ordenado
}

File: motor/domain/commits.py
"""Porte de internal/domain/commits.go."""

from __future__ import annotations

import re

from motor.domain.types import CommitRef

padrao_chamado = re.compile(r"\bch(\d+)\b")


def match_exato(candidatos: list[CommitRef], chamado: str) -> list[CommitRef]:
    """Filtra candidatos de grep por word-boundary: `ch5514` nao pode casar
    dentro de `ch255514`. search_commits do GitRepo so traz candidatos brutos.
    """
    if not chamado:
        return []
    padrao = re.compile(r"\bch" + re.escape(chamado) + r"\b")
    return [c for c in candidatos if padrao.search(c.msg)]


def extrair_chamado(msg: str) -> str | None:
    """Acha o numero do chamado na mensagem. Usado na reconstrucao do estado,
    para reagrupar por chamado a partir do trailer de cherry-pick.
    """
    m = padrao_chamado.search(msg)
    return None if m is None else m.group(1)


def agrupar_por_chamado(commits: list[CommitRef]) -> dict[str, list[CommitRef]]:
    """Agrupa preservando a ordem de 1a aparicao de cada chamado.

    Commit sem chamado (orfao) cai numa chave propria pelo hash curto, em vez de
    todos se juntarem numa cesta `""`: o CLI e a TUI listam essas chaves como
    titulo, e uma cesta unica esconderia quantos orfaos distintos existem.
    """
    grupos: dict[str, list[CommitRef]] = {}
    for c in commits:
        grupos.setdefault(c.chamado or c.hash_origem[:8], []).append(c)
    return grupos


def ordenar_por_data(commits: list[CommitRef]) -> list[CommitRef]:
    """Ordena por commit_date asc - nao depende de flag do git (§5
    "Ordenacao"). Nao muta a lista de entrada.
    """
    return sorted(commits, key=lambda c: c.commit_date)
